src/main/java/com/google/devtools/build/lib/analysis/config/transitions/TransitionFactory.java

```diff
@@ -0,0 +1,44 @@
+// Copyright 2019 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.google.devtools.build.lib.analysis.config.transitions;
+
+import com.google.devtools.build.lib.analysis.config.transitions.TransitionFactory.TransitionFactoryData;
+
+/**
+ * Helper for the types of transitions that are statically declared but must be instantiated for
+ * each use.
+ */
+public interface TransitionFactory<T extends TransitionFactoryData> {
+
+  /** Interface for types of data that a {@link TransitionFactory} can use. */
+  interface TransitionFactoryData {}
+
+  /** Returns a new {@link ConfigurationTransition}, based on the given data. */
+  ConfigurationTransition create(T data);
+
+  /** Returns {@code true} if the result of this {@link TransitionFactory} is a host transition. */
+  default boolean isHost() {
```

Comment:
I don't remember if I mentioned this in the design review, but  I was hoping we could get away without needing all of isHost(), isSplit(), an isFinal() in the final abstraction. One of the things I was trying to do with the introduction of ConfigurationTransition was remove all cases of:

```
if (isSplitTransition())
  ((SplitTransition) transition).split(...)
else:
  ((PatchTransition) transition).patch()
```
we originally had littered throughout the code. I'd love consuming logic to care as little about the type of transition as possible.

For this review I suggest we keep as-is, but let's explore in followups if we can simplify away any of these when they ultimately get used.

Reply:
I absolutely want to get rid of them, these are just here to make the migration possible.

Added a TODO about removing them.

Reply:
Good to hear!